src/Microsoft.AspNet.Razor.Runtime/BufferEntryCollection.cs

```
@@ -0,0 +1,191 @@
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException("count");
+            }
+            if (value.Length - index < count)
+            {
+                throw new ArgumentOutOfRangeException("count");
+            }
+
+            while (count > 0)
+            {
+                // Split large char arrays into 1KB strings.
+                var currentCount = Math.Min(count, MaxCharToStringLength);
+                Add(new string(value, index, currentCount));
+                index += currentCount;
+                count -= currentCount;
+            }
+            _isModified = true;
+        }
+
+        /// <summary>
+        /// Adds an instance of <see cref="BufferEntryCollection"/> to the buffer.
+        /// </summary>
+        /// <param name="buffer">The buffer collection to add.</param>
+        public void Add([NotNull] BufferEntryCollection buffer)
+        {
+            _buffer.Add(buffer.BufferEntries);
+            _isModified = true;
+        }
+
+        public void Clear()
+        {
+            _buffer.Clear();
+            _isModified = true;
+        }
+
+        /// <inheritdoc />
+        public IEnumerator<string> GetEnumerator()
+        {
+            return new BufferEntryEnumerator(_buffer);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        private sealed class BufferEntryEnumerator : IEnumerator<string>
+        {
+            private readonly Stack<IEnumerator<object>> _enumerators = new Stack<IEnumerator<object>>();
+            private readonly List<object> _initialBuffer;
+
+            public BufferEntryEnumerator(List<object> buffer)
+            {
+                _initialBuffer = buffer;
+                Reset();
+            }
+
+            public IEnumerator<object> CurrentEnumerator
+            {
+                get
+                {
+                    return _enumerators.Peek();
```

this should probably return `null` or at least `throw` something more specific than the `InvalidOperationException` we'll see if `MoveNext()` hasn't been called.  somewhat prefer `null`.